Add a default TTL when caching

stefanvermaas opened the pull request:
Hi there, 

Thanks for the awesome work on the Geocoder gem! We noticed there isn't a default TTL for any cache key after we saw our Redis instances get filled up with cached Geocoder lookup calls.

Based on the current documentation, I would have expected a default TTL of two days. However, that's not the case. To make sure others won't have the same issue, I thought it would make sense to add a default TTL for every key of two days. 

Looking forward to any feedback.

stefanvermaas commented:
> Thanks for the kind words, and the contribution. 

🙌 

> This is definitely worth discussing. I agree that the README is confusing (and I like your re-ordering of set_defaults), however, I'm not sure that introducing a default TTL is a good idea. For one thing, it may be surprising to people with existing caches that are full of data.

Hmm, that's indeed a fair concern. I didn't consider that. What if we add a deprecation warning first - before merging this OR setting a default TTL - and after that introduce a breaking change version after it introduces a default TTL. 

I noticed some other "deprecation" warnings too. Might be worth it. However, we will first need to find something about your other remark to be fair :)! 

> I've actually been wondering what the use cases are for cache expiration. Surely it varies depending on what you're geocoding (IP address locations change often but physical addresses rarely), what kind of quota/speed your API has, and how quickly your app needs to respond to requests. Since needs vary so widely I lean towards making the default as innocuous/non-destructive as possible. What do you think?

Yes, I can see what you're saying. Our problem is that there are so many possible addresses that it will fill up a Redis instance pretty quickly. For example, we had 110K addresses in Redis being cached and we couldn't push any new ones as the instance was full. I know that hardware is _relatively_ cheap these days, but I like to be conscious about it. 

Thereby I'm not sure if caching all possible addresses in - for us - the US would make any sense. I mean, I've got no problems with calling an API as long as it doesn't have to happen every single time. 

That's why we're using caching. To make sure we're not hitting that API every single time. I'm personally fine with hitting the API once a week. If it's not being hit anymore for that specific address, I would love to see it free up space for new cache keys. 

I mean; why keep space allocated for a key that's only touched once a week/month/year? I'm happy to "pay" for the extra API call. At least, that's our reasoning. 

Love to hear your opinion!

alexreisner commented:
Very interesting. Thanks for describing how you're using the cache. If the main problem is the cache filling up, however, then another solution would be to [configure Redis as an LRU cache](https://redis.io/topics/lru-cache) and evict the oldest/least used entries when new ones are added. That will keep the cache full of as much useful data as possible.

I'm still hesitant to change the default because a default should be the right choice for the majority of apps out there. If use cases vary quite widely (which I suspect they do in this case), then the default should be whatever is least surprising, which to me still seems like no expiration.

stefanvermaas commented:
Yeah, you’re right there. We should set it up as a LRU.

I also get your hesitation. What I do think that would be helpful is to have some more documentation about this topic. 

Shall I rework the PR to enhance the documentation instead? It would allow others to make a better informed decision.

alexreisner commented:
Yes, I think that makes sense. I actually like the changes you've already made to `README.rb` and `configuration.rb`. They clarify what the default values are (and sensibly reorganize the parameters). Can we just roll back the 2-day default expiration?

stefanvermaas commented:
I've rebased the changes, removed the commit that actually set the default to two days, improved the test cases for the default configuration, and updated the documentation. 

What do you think?
